### PR TITLE
virtio_serial_file_transfer: Optimize test case

### PR DESCRIPTION
**Pull request by vivianQizhu**
1. Support tcp socket
2. Support Windows guest better
3. Make it modulized and multiplexible

Depends on https://github.com/avocado-framework/avocado-vt/pull/2135
id: 1724987, 1231039
Signed-off-by: Qianqian Zhu <qizhu@redhat.com>

### CODE REVIEW

**Comment by vivianQizhu** [comment]
@lijinlijin @nanliu-r Please help review, thanks.

**Comment by lijinlijin** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: ERROR: local variable 'host_thread' referenced before assignment (103.31 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: ERROR: local variable 'host_thread' referenced before assignment (84.98 s)

**Comment by vivianQizhu** [comment]
@lijinlijin Updated, please help try again, thanks.

**Comment by vivianQizhu** [comment]
tcp_socket passed on RHEL7 host(only linux pass, windows fail), but fail on RHEL8 host(both linux and windows), I am debugging, will update later.

**Comment by lijinlijin** [comment]
linux guest should failed on rhel7 as well as bug1714393 is a common bz for both windows and linux.

**Comment by vivianQizhu** [comment]
@lijinlijin Maybe because I am using qemu-kvm-1.5.3-167.el7.x86_64, 
 (1/2) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (84.78 s)
 (2/2) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (60.69 s)
But I read through the qemu bug and seems RHEL guest works well? And in my memory linux guest always works well in my test.

**Comment by lijinlijin** [comment]
@vivianQizhu 
Yes, qemu-kvm seems does not have this issue.
For qemu-kvm-rhev, both windows and linux hit this issue(https://bugzilla.redhat.com/show_bug.cgi?id=1714393#c6)

**Comment by vivianQizhu** [comment]
@lijinlijin The script you use to report the qemu bz, will never pass for linux guest, cause it has same issue with https://github.com/avocado-framework/avocado-vt/pull/2135/commits/dd072d6fcf5aeed423a02de4031390753c077e7a and https://github.com/avocado-framework/avocado-vt/pull/2135/commits/337d4f05bc1af417e0d989da6c98cafc064922ae. 
On the new RHEL8(wherer we suppose the bz been fixed), I tested windows guest and it pass(means the issue did fix), RHEL guest still can not receive anything(because `os.read` wont wait for host send anything).
After add the fixes for the script like the above 2 commits, RHEL guest will pass the test.
And with the fixes the RHEL guest can also pass on RHEL7.7 host(where windows guest will fail to receive anything with both your script and this patch).
So my conclusion is, the qemu issue is windows only, and both your script and this patch can detect it. Thanks.

**Comment by lijinlijin** [comment]
@vivianQizhu 
Could you check https://bugzilla.redhat.com/show_bug.cgi?id=1166031#c31
It says rhel guest can be verified on fixed qemu version with the unchanged script.

**Comment by vivianQizhu** [comment]
> @vivianQizhu
> Could you check https://bugzilla.redhat.com/show_bug.cgi?id=1166031#c31
> It says rhel guest can be verified on fixed qemu version with the unchanged script.
@lijinlijin 
The only reason for me to believe this result is that earlier on RHEL guest `os.read` will not loss data after host send process exit. No way for RHEL7.7 and RHEL8.1 guest pass the script, it does not pass the avocado-vt scripts before my fix.
Anyway I would request you try RHEL8.1 guest on latest RHEL8.1 host, it should pass if the script is okay but it does not. Thanks.

**Comment by vivianQizhu** [comment]
@lijinlijin Besides, no matter which version RHEL guest, follow the bz description(comment 0), start guest script first, it will quit immediately.
```
   def _read_header(self):
        if self.is_windows:
            hdr = self._vport.read(READ_HEADER_LEN)     <--------- here windows_support.py has a loop to wait until it read READ_HEADER_LEN data, so windows guest can wait for host send something.
        else:
            hdr = os.read(self._vport, (READ_HEADER_LEN))   <-------- here will read nothing so the size will be 0, nothing will be read. So guest script can never wait for anything being received.
```
So I don't believe follow the bz step RHEL guest will pass your script.

**Comment by lijinlijin** [comment]
@vivianQizhu 
There might be sth wrong during that verification.
But anyway,  your latest result does make sense, I'd like to accept it.
Thanks for the investigate.

File: qemu/tests/virtio_serial_file_transfer.py [hunk 1]
```diff
@@ -1,26 +1,222 @@
...
+    :param file_size: the size of the file to be created
+    :param session: guest session if have one, perform on host if None
+    :return: the full file path
+    """
+    data_file = os.path.join(dir_name,
+                             "tmp-%s" % utils_misc.generate_random_string(8))
+    cmd = "dd if=/dev/zero of=%s bs=1M count=%d" % (data_file, int(file_size))
+    if not session:
+        error_context.context(
+            "Creating %dMB file on guest" % file_size, logging.info)
```
Comment (lijinlijin) [inline qemu/tests/virtio_serial_file_transfer.py]: I guess here should be "Creating %dMB file on host"

File: qemu/tests/virtio_serial_file_transfer.py [hunk 2]
```diff
@@ -1,26 +1,222 @@
...
+        process.run(cmd)
+    else:
+        error_context.context(
+            "Creating %dMB file on host" % file_size, logging.info)
```
Comment (lijinlijin) [inline qemu/tests/virtio_serial_file_transfer.py]: should be "Creating %dMB file on guest"

File: qemu/tests/cfg/virtio_serial_file_transfer.cfg [hunk 3]
```diff
@@ -2,12 +2,11 @@
...
     guest_scripts = VirtIoChannel_guest_send_receive.py;windows_support.py
     guest_script_folder = /var/tmp/
-    file_transfer_serial_port = vc1
-    serials += " vc1 vc2"
-    serial_type_vc1 = virtserialport
-    serial_type_vc2 = virtserialport
+    file_transfer_serial_port = vs1
+    serials += " vs1 vs2"
+    serial_type_vs1 = virtserialport
+    serial_type_vs2 = virtserialport
```
Comment (lijinlijin) [inline qemu/tests/cfg/virtio_serial_file_transfer.cfg]: seems data transfer only use vs1,  so why still add vs2?

**Comment by vivianQizhu** [comment]
@lijinlijin Thanks for review, Updated both this one and https://github.com/avocado-framework/avocado-vt/pull/2135
 (1/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (41.88 s)
 (2/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (40.60 s)
 (3/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (106.84 s)
 (4/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (86.67 s)

**Comment by vivianQizhu** [comment]
@lijinlijin Besides, now I am sure that your scripts has same issue with here: https://github.com/avocado-framework/avocado-vt/pull/2135/commits/36ca9e354f32d5ad01f70cb2031030b25151a4bb
After apply similar fix, windows guest actually can receive data from host with 'the buggy qemu', it seems not receive data just because host process exit before guest receive data. This is not 100% reproducible, I guess some version host easily to trigger this issue and some does not, so mislead the tester like some qemu version is buggy while some qemu version is fixed.

**Comment by vivianQizhu** [comment]
@lijinlijin And another thing is, how easily the bug to be triggered, highly depends on the speed host send data. If we set a small send buffer for host socket, it will send data slower, then the guest almost always can receive the data properly. I guess that might explain why you hit the issue more with unix socket, that might because tcp socket send data slower so that it might wait for guest receiving data, so it pass.

**Comment by lijinlijin** [comment]
@vivianQizhu 
sometimes I hit "ERROR: argument of type 'NoneType' is not iterable" on windows guests:
I print the log, guest and host output both are empty:
 (001/203) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (939.40 s)
 (002/203) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.rh_kernel_update: PASS (453.47 s)
 (003/203) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: -
PASS (1143.54 s)
 (004/203) repeat1.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (286.88 s)
 (005/203) repeat1.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (66.19 s)
 (006/203) repeat1.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (161.98 s)
 (007/203) repeat1.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (149.03 s)
 (008/203) repeat2.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (127.01 s)
 (009/203) repeat2.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (64.55 s)
 (010/203) repeat2.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: ERROR: argument of type 'NoneType' is not iterable (307.94 s)
 (011/203) repeat2.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: ERROR: argument of type 'NoneType' is not iterable (134.95 s)
 (012/203) repeat3.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (282.71 s)
 (013/203) repeat3.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (70.26 s)
 (014/203) repeat3.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: ERROR: argument of type 'NoneType' is not iterable (307.68 s)
 (015/203) repeat3.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: ERROR: argument of type 'NoneType' is not iterable (128.71 s)
 (016/203) repeat4.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (283.29 s)
 (017/203) repeat4.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (70.93 s)
 (018/203) repeat4.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: ERROR: argument of type 'NoneType' is not iterable (308.56 s)
 (019/203) repeat4.Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: ERROR: argument of type 'NoneType' is not iterable (127.64 s)

And from what I observed until now, seems once this issue happened, all repeated virtio_serial_file_transfer will fail unless I reinstall a new image.

**Comment by vivianQizhu** [comment]
@lijinlijin Thanks for the test, I am trying to reproduce it. Would you help upload the complete log for the first fail case to bz 1724987? Thanks.

**Comment by vivianQizhu** [comment]
@lijinlijin Reproduced with Win10 guest, win2019 always works well in my test.
`python` sometimes can not be found in win10, replace it with `python2.7`. Test passed on win10 and win2019(but not repeat so much time).
Just not sure if all windows guest use `python2.7`, we might specify it for specific guest if it use different version. Please help review, thanks.

**Comment by lijinlijin** [comment]
All guests install python2.7 during guest installation:
cmd /c msiexec /package %drive%ActivePython-2.7.8.10-win64-x64.msi /passive INSTALLDIR=%pydir%
There are three python application under c:\Python27: python, python2, python2.7
I check on win10-64, only python does not work.

I run on win2008-32,win7-32, win8-32, win8.1-32 and win10-32 and repeat 5 times, all passed:
Summary: 
Finshed=55, PASS=55

File: qemu/tests/virtio_serial_file_transfer.py [hunk 4]
```diff
@@ -1,26 +1,222 @@
...
+                    err += "match. Script output:\n%s" % output
+                    return False, err
+            else:
+                md5_re = "md5_sum = (\\w{32})"
+                try:
+                    md5_host = re.findall(md5_re, output)[0]
+                except Exception:
+                    err = "Fail to get md5, script may fail."
+                    err += " Script output:\n%s" % output
+                    return False, err
```
Comment (lijinlijin) [inline qemu/tests/virtio_serial_file_transfer.py]: line 111 to line123 and line 126 to line 139 are quite similar, could we make them to a new function?

Comment (vivianQizhu) [reply qemu/tests/virtio_serial_file_transfer.py]: @lijinlijin Updated, please help check, thanks.

**Comment by lijinlijin** [comment]
Acked-by: Li Jin lijin@redhat.com

 (1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.virtio_serial_file_transfer.unix_socket: PASS (122.00 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.virtio_serial_file_transfer.tcp_socket: PASS (134.41 s)

**Pull request merged**